rasa_sdk/interfaces.py

```diff
@@ -247,17 +247,21 @@ def slots_to_validate(self) -> Dict[Text, Any]:
         """
 
         slots: Dict[Text, Any] = {}
+        count: int = 0
 
         for event in reversed(self.events):
             # The `FormAction` in Rasa Open Source will append all slot candidates
             # at the end of the tracker events.
             if event["event"] == "slot":
-                slots[event["name"]] = event["value"]
+                count += 1
             else:
                 # Stop as soon as there is another event type as this means that we
                 # checked all potential slot candidates.
                 break
 
+        for event in self.events[len(self.events)-count:]:
```

Comment:
the change makes sense to me - just wondering if we can find a more elegant way to do this (e.g. using `enumerate`). Especially the name `count` in my opinion hides the fact that we are dealing with indexes here.

Reply:
"count" is indeed the count of slot events but I can understand your concern given it is later used to compute index (but same would go for len() which can often be used in index calculation also). I agreed enumerate would be more elegant but I was not confident enough to make such change.

What if we simply iterate through self.events entirely in its original order at the first place? I am not familiar with the internals enough to gauge the impact of the loss performance but the code should be easier to read and free from the assumption of slot events being last.

Reply:
> What if we simply iterate through self.events entirely in its original order at the first place?

Mhm, there could be a lot of events from the conversation (I guess something in the range from 0 to 100) while we can be sure that the information we need (the `SlotSet` events) are at the end of this list.

> free from the assumption of slot events being last.

That assumption is actually crucial. Otherwise there is no way to distinguish whether the slots are set by the `FormAction` within Rasa Open Source or whether they were set earlier in the conversation.

Reply:
Thanks @wochinge for the explanation.